vexim/config/functions.php

```diff
@@ -99,39 +99,37 @@ function alpha_menu($flag)
     function crypt_password($clear, $salt = '')
     {
         global $cryptscheme;
-        
-        switch($cryptscheme){
-                case 'sha':
-                    $hash = sha1($clear);
-                    $cryptedpass = '{SHA}' . base64_encode(pack('H*', $hash));
-                break;
- 		case 'des':
-                    if (!empty($salt))
-                    {
-                        $salt = substr($salt, 0, 2);
-                    }
-                    else
-                    {
-                        $salt = get_random_bytes(2);
-                    }
-                    $cryptedpass = crypt($clear, $salt);
+
+        if($cryptscheme === 'sha') {
+          $hash = sha1($clear);
+          $cryptedpass = '{SHA}' . base64_encode(pack('H*', $hash));
+        } elseif ($cryptscheme === 'clear') {
+          $cryptedpass=$clear;
+        } else {        
+          if(empty($salt)) {
+            switch($cryptscheme){
+              case 'des':
+                $salt = '';
+                break;            
+              case 'md5':
+                $salt='$1$';
                 break;
-                case 'md5':
-                    if (!empty($salt))
-                    {
-                        $salt = substr($salt, 0, 12);
-                    }
-                    else
-                    {
-                        $salt = '$1$'.get_random_bytes(8).'$';
-                    }
-                    $cryptedpass = crypt($clear, $salt);
+              case 'sha512':
+                $salt='$6$';
                 break;
-		case 'clear':
-                    $cryptedpass=$clear;
+              case 'bcrypt':
+                $salt='$2a$10$';
                 break;
-		default:
-		   $cryptedpass = crypt($clear, $salt);
+              default:
+                if(preg_match('/\$[:digit:][:alnum:]?\$/', $cryptscheme)) {
+                  $salt=$cryptscheme;
+                } else {
+                  die(_('The value of $cryptscheme is invalid!'));
+                }
+            }
+           $salt.=get_random_bytes(CRYPT_SALT_LENGTH);
```

Comment:
Huh, i think I know why Avleen appended the `$` here: because of [this](http://man7.org/linux/man-pages/man3/crypt.3.html#NOTES). I wonder if there's a mistake in this manpage, or if this behavior has been changed and this bit of the manual overlooked, or if PHP just works differently... Either way, I guess it might be safer to add it back. I'm sorry for being so inconsistent with my comments.

Also, spacing is still inconsistent in this block.